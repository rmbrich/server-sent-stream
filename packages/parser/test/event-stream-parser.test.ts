import {describe} from '@jest/globals';

import EventStreamParser from '../src/index';
import testParser from '../../../fixtures/parsing';

const testEndToEnd = (chunks: string[]) => {
    const events: {data: string, eventType: string, lastEventId: string}[] = [];
    const parser = new EventStreamParser((data, eventType, lastEventId) => {
        events.push({data, eventType, lastEventId});
    });
    for (const chunk of chunks) {
        parser.push(chunk);
    }
    parser.end();
    return Promise.resolve(events);
};

describe('EventStreamParser', () => {
    testParser(testEndToEnd);
});
