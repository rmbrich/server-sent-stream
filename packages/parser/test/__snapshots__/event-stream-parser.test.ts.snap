// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventStreamParser basic cr 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser basic crlf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser basic lf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at cr followed by crlf 1 1`] = `
[
  {
    "data": "abc",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "def",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at cr followed by crlf 2 1`] = `
[
  {
    "data": "abc",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "def",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at crlf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at end cr 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at end crlf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at end lf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at space cr 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at space crlf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunk split at space lf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunked cr 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunked crlf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser chunked lf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser comment cr 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser comment crlf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser comment lf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser cr followed by crlf 1`] = `
[
  {
    "data": "abc",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "def",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser empty event type is treated as 'message' cr 1`] = `
[
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser empty event type is treated as 'message' crlf 1`] = `
[
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser empty event type is treated as 'message' lf 1`] = `
[
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser event ID with null byte cr 1`] = `
[
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "bar",
  },
  {
    "data": "again",
    "eventType": "message",
    "lastEventId": "bar",
  },
  {
    "data": "quux",
    "eventType": "message",
    "lastEventId": "baz",
  },
]
`;

exports[`EventStreamParser event ID with null byte crlf 1`] = `
[
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "bar",
  },
  {
    "data": "again",
    "eventType": "message",
    "lastEventId": "bar",
  },
  {
    "data": "quux",
    "eventType": "message",
    "lastEventId": "baz",
  },
]
`;

exports[`EventStreamParser event ID with null byte lf 1`] = `
[
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "bar",
  },
  {
    "data": "again",
    "eventType": "message",
    "lastEventId": "bar",
  },
  {
    "data": "quux",
    "eventType": "message",
    "lastEventId": "baz",
  },
]
`;

exports[`EventStreamParser event IDs cr 1`] = `
[
  {
    "data": "foo
quux",
    "eventType": "message",
    "lastEventId": "baz",
  },
  {
    "data": "sameid",
    "eventType": "message",
    "lastEventId": "baz",
  },
  {
    "data": "replacedid",
    "eventType": "message",
    "lastEventId": "newid",
  },
]
`;

exports[`EventStreamParser event IDs crlf 1`] = `
[
  {
    "data": "foo
quux",
    "eventType": "message",
    "lastEventId": "baz",
  },
  {
    "data": "sameid",
    "eventType": "message",
    "lastEventId": "baz",
  },
  {
    "data": "replacedid",
    "eventType": "message",
    "lastEventId": "newid",
  },
]
`;

exports[`EventStreamParser event IDs lf 1`] = `
[
  {
    "data": "foo
quux",
    "eventType": "message",
    "lastEventId": "baz",
  },
  {
    "data": "sameid",
    "eventType": "message",
    "lastEventId": "baz",
  },
  {
    "data": "replacedid",
    "eventType": "message",
    "lastEventId": "newid",
  },
]
`;

exports[`EventStreamParser event types cr 1`] = `
[
  {
    "data": "foo
quux",
    "eventType": "baz",
    "lastEventId": "",
  },
  {
    "data": "noeventtype",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser event types crlf 1`] = `
[
  {
    "data": "foo
quux",
    "eventType": "baz",
    "lastEventId": "",
  },
  {
    "data": "noeventtype",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser event types lf 1`] = `
[
  {
    "data": "foo
quux",
    "eventType": "baz",
    "lastEventId": "",
  },
  {
    "data": "noeventtype",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser field with no colon cr 1`] = `
[
  {
    "data": "abc


foo
",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser field with no colon crlf 1`] = `
[
  {
    "data": "abc


foo
",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser field with no colon lf 1`] = `
[
  {
    "data": "abc


foo
",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser invalid stream cr 1`] = `[]`;

exports[`EventStreamParser invalid stream crlf 1`] = `[]`;

exports[`EventStreamParser invalid stream lf 1`] = `[]`;

exports[`EventStreamParser only one leading space stripped cr 1`] = `
[
  {
    "data": " abc",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser only one leading space stripped crlf 1`] = `
[
  {
    "data": " abc",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser only one leading space stripped lf 1`] = `
[
  {
    "data": " abc",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser only one trailing newline stripped cr 1`] = `
[
  {
    "data": " abc

",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser only one trailing newline stripped crlf 1`] = `
[
  {
    "data": " abc

",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser only one trailing newline stripped lf 1`] = `
[
  {
    "data": " abc

",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser very chunked cr 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser very chunked crlf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;

exports[`EventStreamParser very chunked lf 1`] = `
[
  {
    "data": "abc
def",
    "eventType": "message",
    "lastEventId": "",
  },
  {
    "data": "foo",
    "eventType": "message",
    "lastEventId": "",
  },
]
`;
